### ISSUE TITLE
[Enhancement] Add support for model compilation

### ISSUE THREAD

**Issue by Innixma** [awslabs/autogluon#1742]
AutoGluon-Tabular currently trains many models. Each of these models adhere to the `AbstractModel` API.

Add a new method: `AbstractModel.compile` which can be called after `AbstractModel.fit`. `AbstractModel.compile` will perform an operation which creates a new model artifact that is able to perform inference on new data faster than before compilation.

This must be implemented separately for each model, as the way to compile each model can differ (and some models will not support compiling).

Models in order of priority for adding compilation support:

1.  Random Forest / Extra Trees
2. LightGBM
3. TorchNN
4. FastAINN
5. XGBoost
6. CatBoost

Known packages that could provide CPU compilation support:
1. ONNX
2. TreeLite
3. TVM
4. Intel sklearnex
5. lleaves

Accelerated Models:
1. KNN -> Intel sklearnex
2. Linear -> Intel sklearnex

Extra: Average training time, inference time, and model quality on 1040 datasets as of v0.4.0, which we see that RF/XT/NN are the slowest models during inference by a large margin.

```
           framework  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled
0           CatBoost       0.051178            2300.807620               4.059308
1         LightGBMXT       0.069720             129.668734               5.967802
2            XGBoost       0.094320             134.763325               2.704312
3           LightGBM       0.118972              87.351193               4.118561
4      LightGBMLarge       0.141573             221.466008               6.134407
5   RandomForestEntr       0.139603             126.959287              36.647381
6     NeuralNetTorch       0.156578             799.706151              17.693657
7   RandomForestGini       0.147252              92.553456              36.738299
8    NeuralNetFastAI       0.196004             423.930844              19.812870
9     ExtraTreesEntr       0.173157              66.854415              34.559691
10    ExtraTreesGini       0.176122              65.440872              45.290199
11    KNeighborsDist       0.898552               1.000000               8.141010
12    KNeighborsUnif       0.917718              11.649699               7.620200
```

**Comment by bryanyzhu** [comment]
> Ping also @bryanyzhu @zhiqiangdon . This is related to adding ONNX-graph exporting / TVM exporting support in AutoMM.
> 
> This is a really good proposal! For the compile flag, should we aim at option-1 or option-2?
> 
> ```
> # Option-1
> 
> predictor = XXXPredictor()
> predictor.fit(train_data)
> predictor.compile()
> predictions = predictor.predict()
> 
> # Option-2
> 
> predictor = XXXPredictor()
> predictor.fit(train_data)
> inference_only_predictor = predictor.compile()
> predictions = inference_only_predictor.predict()
> ```
> 
> Option-2 has the advantage that we can get multiple inference_only_predictors for different devices.

I like option-2, the same reason as you mentioned.

**Comment by Innixma** [comment]
@sxjscience I prefer option-2 in the long-term because it has obvious benefits of flexibility and avoids non-reversibly changing the original artifact, however option-1 is easier to implement in regards to AutoGluon's stacking logic without major code changes. (I've already done working PoCs for lleaves and ONNX in <100 lines via option-1, would be a lot more code for option-2).

For short-term, we can simplify by saying that we always replace model with its compiled version on disk and return the compiled version, while internally simply updating the model artifact itself, then later we can remove the assumption that we replace the model artifact and can instead save compiled model to a new location and treat them as separate models.

One hack to get option-2 with less code is to link it to Predictor.clone functionality (#1755) and simply clone the predictor and perform option-1 on the clone (effectively the same as doing option-2 on the original)
